### PR TITLE
Tool 111 rewrite networking

### PR DESCRIPTION
**Pull request by BirmacherAkos**
- bundleID: ensure app
- capability: enable, update capability
- Apple service: compare apple service status between the target and the bundleID on the dev. portal

### CODE REVIEW

File: autoprovision/bundleidhelper.go [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+// EnsureApp search for AppID on the developer portal for the provided target's bundleID.
+// If the target is not executable (not app, extension or UITest), then it returns nil
+// If the AppID is available in the developer portal, it will return it
+// If it's not, it will generate a new one with a name of:
+// Bitrise {bundleID} {targetID}. `Example: auto_provision.ios-simple-objc + bc7cd9d1cc241639c4457975fefd920f => Bitrise auto provision ios simple objc bc7cd9d1cc241639c4457975fefd920f`
+func EnsureApp(client *appstoreconnect.Client, projectHelper ProjectHelper, target xcodeproj.Target, configurationName string) (*BundleID, error) {
+	// Check only executable targets which need to be registered on the Dev. Portal
+	if !target.IsExecutableProduct() {
```
Comment (BirmacherAkos) [inline autoprovision/bundleidhelper.go]: The UI test will not be executable anymore. 
https://github.com/bitrise-io/xcode-project/pull/35#discussion_r306839104

Do not forget to update it!

File: appstoreconnect/appstoreconnect.go [hunk 2]
```diff
@@ -72,8 +72,8 @@ func (c *Client) ensureSignedToken() (string, error) {
 			return "", fmt.Errorf("failed to cast claim for token")
 		}
 		expiration := time.Unix(int64(claim.Expiration), 0)
-		// token is marked valid if it will not expire in the upcoming 10 sec
-		if expiration.After(time.Now().Add(10 * time.Second)) {
+		// token is marked valid if it will not expire in the upcoming 60 sec
```
Comment (godrei) [inline appstoreconnect/appstoreconnect.go]: based on: https://developer.apple.com/documentation/appstoreconnectapi/generating_tokens_for_api_requests

`You do not need to generate a new token for every API request. To get better performance from the App Store Connect API, reuse the same signed token for up to 20 minutes.`

File: appstoreconnect/bundleids_test.go [hunk 3]
```diff
@@ -16,6 +14,15 @@ func TestProvisioningService_ListBundleIDs(t *testing.T) {
 			name: "Get bundle ID if for com.bitrise.Test-Xcode-Managed",
 			opt: &ListBundleIDsOptions{
 				FilterIdentifier: "com.bitrise.Test-Xcode-Managed",
+				Include:          "bundleIdCapabilities",
+			},
+			wantErr: false,
+		},
+		{
+			name: "Get bundle ID if for com.bitrise.io.testing.firefox",
```
Comment (godrei) [inline appstoreconnect/bundleids_test.go]: why is it a different test case then the above one?

File: appstoreconnect/capabilities_test.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	tests := []struct {
+		name     string
+		bundleID BundleID
+		wantErr  bool
+	}{
+		{
+			name: "Fetch bundleID capabilitios of com.bitrise.io.testing.firefox",
+			bundleID: BundleID{
+				Relationships: struct {
+					Profiles struct {
```
Comment (godrei) [inline appstoreconnect/capabilities_test.go]: you can remove `Profiles` relationship if you do not define it later

File: autoprovision/bundleidhelper.go [hunk 5]
```diff
@@ -0,0 +1,377 @@
+package autoprovision
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/bitrise-io/xcode-project/serialized"
+
```
Comment (godrei) [inline autoprovision/bundleidhelper.go]: some unnecessary new lines

File: autoprovision/bundleidhelper.go [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+// If the AppID is available in the developer portal, it will return it
+// If it's not, it will generate a new one with a name of:
+// Bitrise {bundleID} {targetID}. `Example: auto_provision.ios-simple-objc + bc7cd9d1cc241639c4457975fefd920f => Bitrise auto provision ios simple objc bc7cd9d1cc241639c4457975fefd920f`
+func EnsureApp(client *appstoreconnect.Client, projectHelper ProjectHelper, target xcodeproj.Target, configurationName string) (*BundleID, error) {
+	// Check only executable targets which need to be registered on the Dev. Portal
+	if !target.IsExecutableProduct() {
+		return nil, nil
+	}
+
+	platform := func(projectPlatform Platform) appstoreconnect.BundleIDPlatform {
```
Comment (godrei) [inline autoprovision/bundleidhelper.go]: hmm why did you solve it like this?

```
	var platform appstoreconnect.BundleIDPlatform
	switch projectPlatform {
	case IOS, TVOS:
		platform = appstoreconnect.IOS
	case MacOS:
		platform = appstoreconnect.MacOS
	default:
		return errors.New("unkown platform: %s", projectPlatform)
	}
```

-1 line

File: autoprovision/bundleidhelper.go [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if b != nil {
+		return b, nil
+	}
+	log.Warnf("No AppID was found with bundleID: %s", target.Name)
+
+	// Generate AppID name from the target bundleID and from targetID
+	// auto_provision.ios-simple-objc + bc7cd9d1cc241639c4457975fefd920f => Bitrise auto provision ios simple objc bc7cd9d1cc241639c4457975fefd920f
```
Comment (godrei) [inline autoprovision/bundleidhelper.go]: why is the target id included in the profile name? bundle ID is unique across the whole dev portal team

File: autoprovision/bundleidhelper.go [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+	// Profiles     []appstoreconnect.Profile
+	ID string
+}
+
+// EnsureApp search for AppID on the developer portal for the provided target's bundleID.
+// If the target is not executable (not app, extension or UITest), then it returns nil
+// If the AppID is available in the developer portal, it will return it
+// If it's not, it will generate a new one with a name of:
+// Bitrise {bundleID} {targetID}. `Example: auto_provision.ios-simple-objc + bc7cd9d1cc241639c4457975fefd920f => Bitrise auto provision ios simple objc bc7cd9d1cc241639c4457975fefd920f`
+func EnsureApp(client *appstoreconnect.Client, projectHelper ProjectHelper, target xcodeproj.Target, configurationName string) (*BundleID, error) {
```
Comment (godrei) [inline autoprovision/bundleidhelper.go]: why not just passing the target's bundleID?

File: autoprovision/bundleidhelper.go (hunk 1 above)
Comment (godrei) [inline autoprovision/bundleidhelper.go]: why do you do this check inside this function? i feel like the client-side should decide which targets needs to be signed

File: autoprovision/bundleidhelper.go [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+	for _, cap := range c.Data {
+		bundleIDCapabilities = append(bundleIDCapabilities, cap)
+	}
+	return bundleIDCapabilities, nil
+}
+
+// syncAppServices compares the target's capabilities one-by-one with the AppID's capability list on the developer portal,
+// If the capability is not ethen enables it.
+func syncAppServices(client *appstoreconnect.Client, projectHelper ProjectHelper, target xcodeproj.Target, configurationName string, bundleID BundleID) error {
+	targetEntitlements, err := projectHelper.targetEntitlements(target.Name, configurationName)
```
Comment (godrei) [inline autoprovision/bundleidhelper.go]: why not passing the `targetEntitlements` to this function?
why do you pass the target?

File: autoprovision/bundleidhelper.go [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+		return fmt.Errorf("failed to get target's data procetion entitlement, error: %s", err)
+	} else if targetDataProtectionValue != "" {
+		if err := updateDataProtection(client, bundleID, targetDataProtectionValue); err != nil {
+			return err
+		}
+	}
+
+	//
+	// iCloud
+	usesICloudDocuments, usesICloudKit, usesICloudKeyValueStorage, err := usesICloudServices(targetEntitlements)
```
Comment (godrei) [inline autoprovision/bundleidhelper.go]: previously we had 2 different cases around iCloud services: https://github.com/bitrise-steplib/steps-ios-auto-provision/blob/master/lib/autoprovision/portal/app_client.rb#L181-L193
Here you have only one, do not you miss something?

File: autoprovision/bundleidhelper.go [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+	}
+
+	if len(iCloudServices) > 0 {
+		usesICloudDocuments = sliceutil.IsStringInSlice("CloudDocuments", iCloudServices)
+		usesICloudKit = sliceutil.IsStringInSlice("CloudKit", iCloudServices)
+	}
+	return
+}
+
+func enableAppService(client *appstoreconnect.Client, capabilityType appstoreconnect.CapabilityType, bundleID BundleID, settings []appstoreconnect.CapabilitySetting) error {
```
Comment (godrei) [inline autoprovision/bundleidhelper.go]: why not just the bundleID string is passed?

File: autoprovision/bundleidhelper.go [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+			},
+			Type: "bundleIdCapabilities",
+			ID:   capabilityID,
+		},
+	})
+	return err
+}
+
+func updateICloud(client *appstoreconnect.Client, bundleID BundleID, targetICloudVersion string) error {
+	iCLoudVersion, iCloudCapabilityID := func() (appstoreconnect.CapabilityOptionKey, string) {
```
Comment (godrei) [inline autoprovision/bundleidhelper.go]: again why did you solve it using a function call?

could be:

```
	var iCLoudVersion, iCloudCapabilityID string
	for _, bundleIDCap := range bundleID.Capabilities {
		if bundleIDCap.Attributes.CapabilityType == appstoreconnect.ICloud {
			for _, settings := range bundleIDCap.Attributes.Settings {
				if settings.Key == appstoreconnect.IcloudVersion {
					iCLoudVersion, iCloudCapabilityID = settings.Options[0].Key, bundleIDCap.ID
				}
			}
		}
	}
```

File: autoprovision/bundleidhelper.go [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+	capabilitySettingOption := appstoreconnect.CapabilityOption{Key: xcodeVersion}
+	capabilitySetting := appstoreconnect.CapabilitySetting{
+		Options: []appstoreconnect.CapabilityOption{capabilitySettingOption},
+		Key:     appstoreconnect.IcloudVersion,
+	}
+	return updateAppService(client, capabilityID, appstoreconnect.ICloud, []appstoreconnect.CapabilitySetting{capabilitySetting})
+}
+
+func updateDataProtection(client *appstoreconnect.Client, bundleID BundleID, targetProtectionValue string) error {
+	protectionCapabilityValue, protectionCapabilityID := func() (appstoreconnect.CapabilityOptionKey, string) {
```
Comment (godrei) [inline autoprovision/bundleidhelper.go]: again why did you solve it using a function call?

File: autoprovision/bundleidhelper.go [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+					if settings.Key == appstoreconnect.DataProtectionPermissionLevel {
+						return settings.Options[0].Key, bundleIDCap.ID
+					}
+				}
+			}
+		}
+		return "", ""
+	}()
+
+	switch targetProtectionValue {
```
Comment (godrei) [inline autoprovision/bundleidhelper.go]: should not the possible targetProtectionValues defined as const pr as a map[targetProtectionValue] appstoreconnect.Protection ?

File: autoprovision/bundleidhelper.go [hunk 15]
```diff
@@ -1,116 +1,96 @@
...
-	// Check only executable targets which need to be registered on the Dev. Portal
-	if !target.IsExecutableProduct() {
-		return nil, nil
+func appIDName(bundleID string) string {
+	r := strings.NewReplacer(".", " ", "_", " ", "-", " ")
+	return "Bitrise " + r.Replace(bundleID)
+}
+
+// EnsureApp search for AppID on the developer portal for the provided bundleID.
+// If the AppID is available on the developer portal, it will ne returned.
```
Comment (BirmacherAkos) [inline autoprovision/bundleidhelper.go]: Typo
it will **ne** returned.

**Pull request merged**